### PR TITLE
send popup message to chat

### PR DESCRIPTION
**Pull request by onoira**
## About the PR

display popup messages in the chat

## Why / Balance

in accordance with the information redundancy principle from the accessibility guidelines:
<https://hackmd.io/PIF3dxZkRS27hIyJM58xLw#Text>

the popup itself still provides an animated alert in the play area, but can also be read later or grab the player's attention if they're looking away from the play area, _that is_ at the chat.

## Technical details

adds a `notice` command that writes a client-side message to the `Visual` ('Actions') channel. `Visual` is filtered out by default.

originally I tried sending this only when the popup target was the player, but this excluded things like _burning your hand_ or _getting injected with a needle_ that seem important. rather than enforce or expect an important popup to always centre on the player character, this assumes that any popup is something intended to be read.

this PR is not concerned with the style or consistency of popup messages; only that they get duplicated to the chat.

yes, I see the deprecation warning on `IConsoleCommand`, but I don't know what to do with that information.
yes, this allows people to talk to themselves in the chat window. I don't know if that's good  or bad.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, ~~**or** this PR does not require an ingame showcase~~

image

<details>
<summary>chat filter</summary>

image

</details>

## Breaking changes

n/a

## Changelog

no cl?

### CODE REVIEW

**Comment by lzk228** [comment]
really like it, but maybe apply some text formatting?

**Comment by ike709** [comment]
I think this is chatlog bloat and I'd like a setting to disable it.

**Comment by mirrorcult** [comment]
this should be opt-in. chat-channel thats disabled by default i think

**Comment by lzk228** [comment]
> ready for review

can you update the media please?

File: Content.Server/Chat/Commands/NoticeCommand.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                return;
+
+            var message = string.Join(" ", args).Trim();
+            if (string.IsNullOrEmpty(message))
+                return;
+
+            IoCManager.Resolve<IChatManager>().ChatMessageToOne(ChatChannel.Visual, message, message, EntityUid.Invalid, false, player.ConnectedClient, recordReplay: false);
+        }
+    }
+}
```
Comment (metalgearsloth) [inline Content.Server/Chat/Commands/NoticeCommand.cs:44]: Please update to code standards:
1. localisation
2. file-scoped namespace
3. public
4. Use dependencies not resolve.

File: Content.Client/Popups/PopupSystem.cs [hunk 2]
```diff
@@ -73,6 +76,7 @@ private void PopupMessage(string message, PopupType type, EntityCoordinates coor
             };
 
             _aliveWorldLabels.Add(label);
+            _chatManager.SendMessage($"notice {message}", ChatSelectChannel.Console);
         }
```
Comment (metalgearsloth) [inline Content.Client/Popups/PopupSystem.cs:80]: How does this work with replays.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I don't understand why we need to roundtrip to server for a local chat message.

File: Content.Server/Chat/Commands/NoticeCommand.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/Chat/Commands/NoticeCommand.cs]: Make it a public class not internal please.

File: Content.Server/Chat/Commands/NoticeCommand.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        if (player.Status != SessionStatus.InGame)
+            return;
+
+        if (player.AttachedEntity is not { })
+        {
+            shell.WriteError("You don't have an entity!");
+            return;
+        }
```
Comment (metalgearsloth) [inline Content.Server/Chat/Commands/NoticeCommand.cs:34]: Still needs localising please, everything the client can see needs to be localised.

**Comment by metalgearsloth** [comment]
> > I don't understand why we need to roundtrip to server for a local chat message.
> 
> popup system is clientside; chat system is serverside. we only want the popups the player can see, so we're only treating the popups that should be drawn.

Actually think I'm a dumbass so please ignore, will check ingame later.

**Comment by Arteben** [comment]
Is this pull request still relevant?

**Comment by nok-ko** [comment]
> > Is this pull request still relevant?
> 
> yeah, but i can't presently work on it. i may continue in a few weeks.

Do you still intend to work on this? I would be interested in taking it over if you’re too busy.

**Comment by nok-ko** [comment]
> > Do you still intend to work on this? I would be interested in taking it over if you’re too busy.
> 
> i plan to work on it this weekend, but if you'd prefer to do it for any reason: i'd be okay handing it off.

In that case, go ahead. I might lurk in the code reviews. :)

**Comment by metalgearsloth** [comment]
Re-open if you come back to it.

**Pull request closed without merging**